src/v/archival/segment_reupload.cc

```diff
@@ -0,0 +1,251 @@
+}
+
+model::offset segment_collector::begin_inclusive() const {
+    return _begin_inclusive;
+}
+
+model::offset segment_collector::end_inclusive() const {
+    return _end_inclusive;
+}
+
+const storage::ntp_config* segment_collector::ntp_cfg() const {
+    return _ntp_cfg;
+}
+
+bool segment_collector::can_replace_manifest_segment() const {
+    return _can_replace_manifest_segment && _begin_inclusive < _end_inclusive;
+}
+
+cloud_storage::segment_name segment_collector::adjust_segment_name() const {
+    vassert(
+      !_segments.empty(), "Cannot calculate segment name with no segments");
+
+    auto first = _segments.front();
+    auto file_name = first->filename();
+    auto meta = storage::segment_path::parse_segment_filename(file_name);
+    auto version = meta ? meta->version : storage::record_version_type::v1;
+
+    cloud_storage::segment_name name{};
+    if (_begin_inclusive == first->offsets().base_offset) {
+        auto orig_path = std::filesystem::path(file_name);
+        name = cloud_storage::segment_name(orig_path.filename().string());
+        vlog(archival_log.debug, "Using original segment name: {}", name);
+    } else {
+        auto path = storage::segment_path::make_segment_path(
+          *_ntp_cfg, _begin_inclusive, first->offsets().term, version);
+        name = cloud_storage::segment_name(path.filename().string());
+        vlog(archival_log.debug, "Using adjusted segment name: {}", name);
+    }
+
+    return name;
+}
+
+void segment_collector::align_begin_offset_to_manifest() {
+    if (_begin_inclusive >= _manifest->get_last_offset()) {
+        return;
+    }
+
+    if (_begin_inclusive < _manifest->get_start_offset().value()) {
+        // _begin_inclusive is behind manifest, skip it forward to the start
+        // of manifest.
+        _begin_inclusive = _manifest->get_start_offset().value();
+        vlog(
+          archival_log.debug,
+          "_begin_inclusive is behind manifest for ntp, skipped forward to "
+          "start of manifest from: {} to: {}",
+          _manifest->get_ntp(),
+          _begin_inclusive);
+        return;
+    }
+
+    auto it = _manifest->find(_begin_inclusive);
+
+    // If iterator points to a segment, it means that _begin_inclusive is
+    // aligned on manifest segment boundary, so do nothing. Otherwise, skip
+    // _begin_inclusive to the start of the next manifest segment.
+    if (it == _manifest->end()) {
+        it = _manifest->segment_containing(_begin_inclusive);
+
+        // Pre-conditions make sure that begin_inclusive is inside manifest
+        vassert(it != _manifest->end(), "_begin_inclusive is outside manifest");
+
+        // Skip _begin_inclusive to the start of the next segment. In the corner
+        // case that _begin_inclusive is inside the last segment, it will be
+        // skipped to outside manifest boundary, and the collection will be
+        // skipped.
+        _begin_inclusive = it->second.committed_offset + model::offset{1};
```

Comment:
Actually, the gap is possible because we can disable SI then storage gc can evict some segments then SI can be re-enabled. The code inside archival_policy is prepared for this. If it can't find next segment due to the gap it will bump corresponding metric and continue with the next available segment.

I'm not sure that this is totally correct since this isn't something that we're encouraging users to do. For instance, it will be tricky to consume data from the partition with an offset gap. But also, I'd say that the upload should be as resilient as possible. Because if the upload loop could stuck we will run out of disk space.

Reply:
also, it is possible for the compacted segment to be deleted locally when both compaction and deletion are enabled, so we can upload first compacted segment, then the next will be deleted and we will encounter a gap

Reply:
This requires some more testing but I think it will work with a couple of extra checks and more tests around gap placement.

When we collect compacted segments from local log, if there is a manifest gap somewhere between the aligned offsets, we just re-upload the data ignoring the gap, it will fill the gap in the manifest.

For the gap placed such that the begin or end offset lies inside the gap, the `segment_containing` function will produce a false negative. I have added checks to keep the begin and end offsets as it is if they are detected to be in the gap (rather than aligning them to manifest segments). This will allow the possibility of reducing the gap by uploading data which we have locally but was missed from uploading. It needs a few more unit tests.


> also, it is possible for the compacted segment to be deleted locally when both compaction and deletion are enabled, so we can upload first compacted segment, then the next will be deleted and we will encounter a gap

If I understand correctly, for this case where the local compacted segment is deleted from disk before being uploaded, the log start will have advanced ahead of the last uploaded compacted offset. 

The `find_next_compacted_segment` method skips forward to log start if the search offset is less than that, so it should take care of that scenario. This scenario is tested in `test_segment_collection` where the log starts at offset 5 but we start collection at offset 4.

Reply:
we're using the same approach in remote_partition, it also uses upper_bound and steps back

Reply:
With the latest changes we try to handle gaps, with the approach that if it is possible to fill in the gap in manifest during re-upload then we try to do so. In the scenario where SI is switched off then on, re-upload has a good chance to fill the gap because the data may still be in local disk log which is missing in SI. 

Still need to test a few corner cases, because with gaps and re-upload there are a few more scenarios which could happen.